### PR TITLE
Vasil features: add PlutusV2 support

### LINKED ISSUE #682: Vasil features: add PlutusV2 support
Follow these changes in CSL: https://github.com/Emurgo/cardano-serialization-lib/pull/458

Modify our (de)serialization and domain types accordingly

Part of https://github.com/Plutonomicon/cardano-transaction-lib/issues/567

### PR DESCRIPTION
**Pull request by jy14898**
Closes #682.

### Pre-review checklist

- [x] All code has been formatted using our config (`make format` for Purescript, `nixpkgs-fmt` for Nix)
- [x] All Purescript imports are explicit, including constructors
- [ ] **All** existing examples have been run locally against a fully-synced testnet node
- [x] The integration and unit tests have been run (`npm run test`) locally
- [x] The changelog has been updated under the `## Unreleased` header, using the appropriate sub-headings (`### Added`, `### Removed`, `### Fixed`)

### CODE REVIEW

File: src/Hashing.js [hunk 1]
```diff
@@ -47,7 +47,9 @@ exports.hashPlutusData = plutusData => {
   return lib.hash_plutus_data(plutusData).to_bytes();
 };
 
-exports.hashPlutusScript = plutusScriptBytes => () => {
+exports.hashPlutusScript = getBytes => plutusScript => () => {
+  const plutusScriptBytes = getBytes(plutusScript);
+  // TODO Check if plutusv2 needs to be treated differently
```
Comment (jy14898) [inline src/Hashing.js:52]: Indeed I do need to prefix plutus v2 with a different value

File: test/Transaction.purs [hunk 2]
```diff
@@ -112,7 +112,7 @@ testSetScriptDataHash = liftEffect $ do
       -- Verify the hash with some external tool
       byteArrayToHex sdh
         `shouldEqual`
-          "3ed3d611bc67ef89de1ef8200e4af38210be6c1cfa436e2fef90c7ad48a33df9"
+          "022b4e6f5e29a81f1a84afcc9b0b1487e719ffd95aad04149080dfa74483d4f4"
```
Comment (jy14898) [inline test/Transaction.purs:115]: Is this the right thing to do? With the addition of the plutusv2 cost model, the hash changes

File: test/Fixtures/CostModels.purs [hunk 3]
```diff
@@ -9,179 +9,182 @@ import Data.Tuple.Nested ((/\))
...
-          , 32
-          , 150000
-          , 32
-          , 3345831
-          , 1
-          , 1
-          ]
-      )
+  [ PlutusV1 /\ costModelFixture
+  , PlutusV2 /\ costModelFixture
```
Comment (ngua) [inline test/Fixtures/CostModels.purs:189]: `PlutusV2` has different cost models

Comment (jy14898) [reply test/Fixtures/CostModels.purs:189]: I've done most of the work to support PlutusV2 cost models, however I've hit a CSL bug: https://github.com/Emurgo/cardano-serialization-lib/issues/473

Comment (jy14898) [reply test/Fixtures/CostModels.purs:189]: I think I might still be able to work with it via the bytes interface, however. I'm going to give it a go

Comment (jy14898) [reply test/Fixtures/CostModels.purs]: @ngua Should we update the plutusv1 cost model fixture to the vasil values?

Comment (ngua) [reply test/Fixtures/CostModels.purs]: Thanks @vsubhuman!

@jy14898 I think we should just remove the PS fixture and use CSL's `TxBuilderConstants.plutus_vasil_cost_models`. We can convert it to our `Costmdls` with `Deserialization.Transaction.convertCostModels`

File: src/QueryM/Ogmios.purs [hunk 4]
```diff
@@ -811,177 +816,31 @@ type CostModel =
...
-  , model."verifyEd25519Signature-memory-arguments"
-  ]
+-- This assumes that cost models are stored in lexicographical order
+convertCostModel
+  :: forall costModel
+   . HFoldl (Array Int -> Int -> Array Int) (Array Int) costModel (Array Int)
+  => costModel
+  -> T.CostModel
+convertCostModel model = wrap $ reverse $ hfoldl
+  ((\xs x -> x : xs) :: Array Int -> Int -> Array Int)
```
Comment (klntsky) [inline src/QueryM/Ogmios.purs:841]: Please use `List` and convert to array at the end, `:` on `Array` is `O(n)`.

File: CHANGELOG.md [hunk 5]
```diff
@@ -20,6 +21,9 @@ The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/)
 
 ## [Unreleased]
 
+### Changed
+- `PlutusScript` is now aware of which version of Plutus the script is for. The JSON representation has thus changed to reflect this and is not compatible with previously compatible JSON.
+
```
Comment (jy14898) [inline CHANGELOG.md:26]: Should this be placed under the 2.0.0-alpha section? I'm not sure if the alpha section is preserved in later releases, or if we simply drop the `-alpha` suffix

**Comment by jy14898** [comment]
This PR doesn't include PlutusV2 tests, I think that's possibly a much larger task which would need multiple PRs if issues are found

**Comment by klntsky** [comment]
> I think that's possibly a much larger task

Why? I'd like to be sure that we keep the code hot in the tests, could you please add a plutusv2 script in one of the fixtures we use for serialization roundtrip tests. CSL does not parse the script contents, so you can put any hexstring I believe.

File: src/Serialization/WitnessSet.purs [hunk 6]
```diff
@@ -159,7 +159,9 @@ convertBootstrap { vkey, signature, chainCode, attributes } = do
   newBootstrapWitness vkey' signature' chainCode attributes
 
 convertPlutusScript :: S.PlutusScript -> Effect PlutusScript
-convertPlutusScript (S.PlutusScript bytes) = newPlutusScript bytes
+convertPlutusScript (S.PlutusScript (bytes /\ language)) = case language of
+  S.PlutusV1 -> newPlutusScript bytes
```
Comment (klntsky) [inline src/Serialization/WitnessSet.purs:163]:
```suggestion
  S.PlutusV1 -> newPlutusV1Script bytes
```

**Review by klntsky — CHANGES_REQUESTED** [review]

**Comment by jy14898** [comment]
> > I think that's possibly a much larger task
> 
> Why? ...

For serialization/deserialization, I can do that. I mostly mean for submission, in combination with the other CIP additions. I think those are out of scope of the original issue anyway actually :+1:

**Comment by jy14898** [comment]
Leaving this as a note for myself: use the `len()` method when unpacking cost models instead of looping until error

**Review by klntsky — CHANGES_REQUESTED** [review]

File: CHANGELOG.md [hunk 7]
```diff
@@ -20,7 +21,8 @@ The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/)
 
 ## [Unreleased]
 
-## Changed
+### Changed
+- `PlutusScript` is now aware of which version of Plutus the script is for. The JSON representation has thus changed to reflect this and is not compatible with previously compatible JSON.
```
Comment (klntsky) [inline CHANGELOG.md:25]:
```suggestion
- `PlutusScript` is now aware of which version of Plutus the script is for. The JSON representation has thus changed to reflect this and is not compatible with older JSON format.
```

**Review by klntsky — CHANGES_REQUESTED** [review]

**Review by klntsky — APPROVED** [review]

**Pull request merged**